synchronization/futex.rs

```diff
@@ -0,0 +1,524 @@
+    /// data in the futex.
+    fn fetch_add(&self, val: i32, ordering: Ordering) -> i32;
+}
+
+/// The base Futex type. It literally has to be a 4-byte aligned integer-width piece of
+/// memory and that is all.
+///
+/// Futexes are, from a synchronization stanadpoint, used as "events". This means that
+/// they can be set or unset and can be set/unset/waited on from any other task as well.
+/// They will wake up all listeners when woken.
+///
+/// This implementation of Futex locking is not robust against an owner dying.
+///
+/// When constructing this in shared memory or whatever, something along the lines of:
+/// ```rust
+/// # use futex::Futex;
+/// # use futex::RawFutex;
+/// # let shmpointer: *const i32 = &8;
+/// let shmfutex: &Futex = unsafe {
+///     & *(shmpointer as *const i32 as *const Futex)
+/// };
+/// shmfutex.lock_pi(1, None);
+/// ```
+///
+/// Other than that the current option is to use `std::mem::zeroed`.
+/// Turns out Linux primitives do not pretty Rust make.
+#[repr(align(4), C)]
+pub struct Futex(AtomicI32);
+
+/// Wrapper around a Futex allowing common mutex operations.
+pub struct FutexMutex<T: RawFutex>(T);
+
+#[repr(i32)]
+#[derive(Debug)]
+#[allow(dead_code)]
+enum FutexOp {
+    CmpRequeue = libc::FUTEX_CMP_REQUEUE,
+    CmpRequeuePi = libc::FUTEX_CMP_REQUEUE_PI,
+    Fd = libc::FUTEX_FD,
+    LockPi = libc::FUTEX_LOCK_PI,
+    Requeue = libc::FUTEX_REQUEUE,
+    TrylockPi = libc::FUTEX_TRYLOCK_PI,
+    UnlockPi = libc::FUTEX_UNLOCK_PI,
+    Wait = libc::FUTEX_WAIT,
+    WaitBitset = libc::FUTEX_WAIT_BITSET,
+    WaitRequeuepi = libc::FUTEX_WAIT_REQUEUE_PI,
+    Wake = libc::FUTEX_WAKE,
+    WakeBitset = libc::FUTEX_WAKE_BITSET,
+    WakeOp = libc::FUTEX_WAKE_OP,
+}
+
+thread_local! {
+    static TID: RefCell<libc::pid_t> = RefCell::new(0);
+}
+
+/// Function called after fork in the child to reset the TID
+extern "C" fn clear_tid_on_fork() {
+    TID.with(|t| *t.borrow_mut() = 0);
+}
+
+/// This installs the hook to clear the TID when fork is called.
+fn install_pthread_hook() {
+    let result = unsafe { libc::pthread_atfork(None, None, Some(clear_tid_on_fork)) };
+
+    errno::Errno::result(result).unwrap_or_else(|_| {
+        panic!(
+            "pthread_atfork(None, None, {:p}) failed",
+            clear_tid_on_fork as *const ()
+        )
+    });
+}
+
+#[inline]
+fn gettid() -> libc::pid_t {
+    if unlikely(TID.with(|t| *t.borrow() == 0)) {
+        initialize_per_thread();
+    }
+
+    TID.with(|t| t.borrow().clone())
+}
+
+fn gettid_impl() -> libc::pid_t {
+    let result = unsafe { syscall(SYS_gettid) as libc::pid_t };
+
+    errno::Errno::result(result).expect("gettid() failed")
+}
+
+static PTHREAD_ONCE: std::sync::Once = std::sync::Once::new();
+
+/// Each thread running this code needs to have its thread local thread ID set,
+/// and we install a hook so that a call to pthread_fork will clear the TID.
+///
+/// The hook should only be installed once per process.
```

Comment:
Would it make sense to add in a guarantee to ensure that this is the case / that it only _can_ be installed once per process?